### DISCUSSION TITLE
`fpm.app` support

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#689] [General]
<details><summary>Status: Partially Done</summary>

Follow: https://github.com/ftd-lang/fpm/discussions/550

Initial PR: https://github.com/ftd-lang/fpm/pull/601
processor `fpm-apps` PR: https://github.com/ftd-lang/fpm/pull/618
can_read on fpm.app: https://github.com/ftd-lang/fpm/pull/619
fpm.app fixes: https://github.com/ftd-lang/fpm/pull/630

## Tasks

- [x] Need to implement multi mount-point
- [x] create one type in fpm.ftd of type `fpm.app`
- [x] parse that to fpm package and sitemap
- [x] Need to handle
  - [x] request url prefix may be start with fpm-app multi-mount point
  - [x] If it starts, than that package sitemap is going to use to serve the ftd document
  - [x] Similar thing is going to happen with endpoint, if urls did not match with
    package sitemap, so send request to endpoint, if exists in the `-- fpm.app`.
  - [x] Have to refactor the redirect logic
- [x] `fpm-apps` processor
- [x] readers and writers in `fpm.app`
- [ ] `identity` header support, processor and endpoint will pass X-FPM_USER-ID to service.
- [ ] rename `mount-point` to `url`
- [ ] remove `mount-point` and `endpoint` from dependency
</details>

When we add a fpm dependency, it may be a regular dependency, or it can be a re-usable app. If it is a re-usable app, we may want to install more than one instance of that app. 

Consider a todo app. We may want to a have a todo app for my family, on amitu.com/family-todos/ and another for my side project: amitu.com/worddb/. I might be collaborating with multiple people in each of those, and I may not people in side project to know about my family for example.

Some SAAS apps have the concept of an organisation or workspace or a team, and they create such boundaries at workspace or team level. But this is work for them to implement. The clean boundaries across org. You have to add extra columns in tables, ensure all queries are filtering out what you can not see etc. 

One way to handle it is to install the app multiple times, and each install has no access to other installs. So if I install the same app on /worddb/ and on /family-todos/, the app will not have to create org concept, and they will both have different database etc, so everything will just work.

We are trying to make the job of app developers easy.

```ftd
-- import: fpm
-- fpm.package: amitu.com


-- fpm.app: Todos for Family
url: /family-todos/
package: fifthtry.github.io/kameri
end-point: <>
config: db_url=$ENV.DB_URL
config: name=Upadhyay Khanadan

-- fpm.app: My Word DB
url: /worddb/
package: fifthtry.github.io/kameri
end-point: <>
config: db_url=$ENV.WORDDB_DB_URL
config: name=WordDB: Database Of Words
```

# Package Containing App

```ftd
-- fpm.app: /foo/
package: amitu.com/my-app/

-- fpm.app: /bar/
package: amitu.com/my-app/
```

Let's not do this. One should create a separate package `amitu.com/my-app/` and use it from `amitu.com` in this case. 

# Dynamic Apps

We can also have  dynamic mount points. The URL captured data will be passed as `config: role=family` if request to `/todos/family/` comes. All other configs would be shared across all possible values of `role`.

```ftd
-- import: fpm
-- fpm.package: amitu.com


-- fpm.app: Todos for Family
url: /todos/<string:role>/
package: fifthtry.github.io/kameri
end-point: <>
config: db_url=$ENV.DB_URL
```

# Group Support In App

If I have a group defined `fifthtry-people` (say based on a people with role `ft-zens` in FifthTry discord server) and I want an app to only be accessible to them:

```ftd
-- import: fpm
-- fpm.package: fifthtry.com


-- fpm.user-group: fifthtry-people
discord-role: 793929082483769345/ft-zens

-- fpm.user-group: fifthtry-admin
discord-role: 793929082483769345/ft-admins

-- fpm.app: Todos for FifthTry
url: /fifthtry-todos/
package: fifthtry.github.io/kameri
readers: fifthtry-people
writers: fifthtry-admin
```

When `readers/writers` is used, people without the reader role can not see the app. If they visit the URL they will get 404. Further, we will send `X-FPM-IS-READER` and `X-FPM-IS-WRITER` headers to the endpoint.

# User ID Support In App

If all users use a single identity provider, we can pass an identity to the app as well:

```ftd
-- import: fpm
-- fpm.package: fifthtry.com

-- fpm.app: Todos for FifthTry
url: /fifthtry-todos/
package: fifthtry.github.io/kameri
identity: github-id
readers: fifthtry-people
```

Here we have said the user who have access to this to `/fifthtry-todos/` must have a `discord-id` on the server `793929082483769345`. The ID will be hashed (`sha256(<server-id> + ":" + <user-id> + <id or url>)`), sent to endpoint using `FPM-USER-ID` header.

In hashing we have used id or url. Sometimes you will want to change the URL of an app without losing the userids, so you can continue to use the old url as `id` field.

We can combine readers and any single identity query. If `readers` is not specified, the page will be assumed public, and request from pm to endpoint will be allowed. End point can then show public information, if `X-FPM-USER-ID` is not set.

Say if you wanted users to be based on `@fifthtry.com` email domain and you ask them to login via gmail, it could have been:

```ftd
-- fpm.app: Todos for FifthTry
url: /fifthtry-todos/
package: fifthtry.github.io/kameri
identity: email
readers: fifthtry-people
```

`fpm` will take care of authentication part, and end-point will get just an `X-FPM-USER-ID` header.

## A Note On Identity

We used `identity: email` and `identity: github-id`. These are static [identity](https://fpm.dev/dev/user-identity/), static because they do not change for a given provider. Every time you login with github, your `github-id` would be the same.

**Comment by AbrarNitk** [comment]
As we can install multiple fpm applications in an fpm package.
While installing the application we have to use a `dependency` and `mount-point`, We can also install fpm apps with the same dependency but with a different `mount-point`.

**Reply by AbrarNitk** [reply]
@amitu 

Should we keep the application name as `caption` not and `mount-point`?

```
-- fpm.app: Todos for Family
mount-point: /family-todos/
package: fifthtry.github.io/kameri
end-point: <>
config: db_url=$ENV.DB_URL
config: name=Upadhyay Khanadan
```

**Reply by AbrarNitk** [reply]
not serving many purposes but just giving a name to the installed app.

**Reply by amitu** [reply]
I love the idea. Go ahead. 

Also keep updating the original description with the latest spec that we are implementing. Comments should be considered throw away stuff.
